zml: sharding

gwenzek opened the pull request:
First rough support for in-process sharding.
The goal of sharding is to allow leveraging several GPUs to collaborate to run one big model.

Current implementation:
* added `platform.sharding()` describing how the set of devices managed by the current process should be used.
* added a `_sharding_info` to shape.
* A tensor / shape axis can be marked as needing sharding using `.withSharding(.{ .a })` or `.withSharding(.{ 1 })`.
* When using `populateModelWeights` tensors with sharded axes, will be populated by sharded buffers.
* when generating the MLIR bytecode for a ZML function the **input** arguments, will see their `_sharding_info` converted into `mhlo.sharding` attributes

gwenzek reviewed (commented):
I've removed the `postInit` thing to automatically shard linear layers.
let's come back on this later

Corendos commented on zml/module.zig:
```diff
@@ -565,47 +644,57 @@ fn assignBlockArguments(v: anytype, block: mlir.Block, start: usize) usize {
     };
-    var ctx: LocalContext = .{
-        .index = 0,
+    var capture: LocalContext = .{
+        .index = start,
```
We should try to be consistent on the name of the `LocalContext` variable

gwenzek replied:
agree, but I'm not a big fan of `local_context` and `inner_context`, was wondering if something like `capture` would make it easier to follow to people coming from languages where "capture" is a thing.

gwenzek replied:
(renamed to context)

Corendos replied:
I've seen the `inner_` prefix used many times, that's why I originally used it. But I get your point

Corendos commented on zml/module.zig:
```diff
@@ -674,56 +765,74 @@ pub fn ExeWithWeights(comptime func: anytype) type {
+            var fba = std.heap.FixedBufferAllocator.init(std.mem.sliceAsBytes(data_buffer));
+            const alloc = fba.allocator();
+
+            const input_buffers = alloc.alloc([*]*pjrt.Buffer, n_devices) catch unreachable;
+            for (input_buffers) |*input| {
+                input.* = (alloc.alloc(*pjrt.Buffer, inner.model_buffer_count + inner.args_buffer_count) catch unreachable).ptr;
+            }
+            fillBuffers(&model, input_buffers, 0, inner.model_buffer_count);
+
+            const output_buffers = alloc.alloc([*]*pjrt.Buffer, n_devices) catch unreachable;
+            for (output_buffers) |*output| {
+                output.* = (alloc.alloc(*pjrt.Buffer, inner.result_buffer_count) catch unreachable).ptr;
+            }
```
Can you make sure we handle these `unreachable` instead ?

Since we are handling the error when allocating `data_buffer`

gwenzek replied:
those are provably unreachable because I've already made the actual allocation line 786.
this is just splitting the pre-existing alloc into smaller slices.
I thought it would be more idiomatic than using pointer arithmetic, but I can also do that if you prefer.

Corendos replied:
I see. I would probably have used slicing because there is no actual allocations, but that's up to you

gwenzek replied:
ok I've rewrote this, let me know what you think

Corendos commented on zml/module.zig:
```diff
@@ -875,6 +990,12 @@ fn computeModuleHash(platform: Platform, module: mlir.Module) u64 {
     const api_version = platform.pjrt_api.version();
     writer.writeInt(i64, api_version.major, .little) catch unreachable;
     writer.writeInt(i64, api_version.minor, .little) catch unreachable;
+    const opts = platform.compilation_options;
+    writer.writeByte(if (opts.sharding_enabled) 1 else 0) catch unreachable;
+    writer.writeByte(opts.sharding_axes.len) catch unreachable;
+    for (opts.sharding_axes.constSlice()) |sharding_ax| {
+        writer.print("{s}", .{sharding_ax}) catch unreachable;
+    }
```
Wouldn't this already appear in the generated StableHLO ?

gwenzek replied:
oops

Corendos reviewed (commented):
Left some minor comments.

I don't totally understand the logic about `contiguous` in `HostBuffer` though.

From my understanding, we might want to handle two cases:
1) The tensor is not sharded in the model file, so it's already contiguous and we should handle sharding when doing the conversion from `HostBuffer` to `Buffer`.
2) The tensor is already sharded in the model file (or multiple model files) and a `HostBuffer` is a collection of contiguous slices. In that case, each shard gets converted to a `Buffer` naively.

In any case, we currently want to support only 1, so we could move the sharding logic to the place where we convert a `HostBuffer` to a `Buffer` ?

gwenzek commented on zml/pjrtx.zig:
```diff
@@ -288,9 +226,9 @@ pub const Event = opaque {
-    pub fn deinit(self: *LoadedExecutable, api: *const Api) void {
-        self.inner().deinit(api);
-    }
+    // pub fn deinit(self: *LoadedExecutable, api: *const Api) void {
+    //     self.inner().deinit(api);
+    // }
```
because it crashes the Zig compiler, same below. didn't have time to investigate.
the "TODO" below refers to both.

gwenzek replied:
(btw I'm tempted to nuke pjrtx.zig)

gwenzek commented:
> I don't totally understand the logic about `contiguous` in `HostBuffer` though.

HostBuffer can have strides, and don't need to be contiguous. This was introduced to support `.pt` where buffer can be stored non-contiguously.
I've also added the `slice1d` utility that allows to slice a `HostBuffer` in several non-contiguous slices.
Since it's possible (and now easy) to create non-contiguous slices, I've added a check in few places that we are indeed working on contiguous `HostBuffer` (which was always the case before adding .pt)

gwenzek commented:
Thanks for the review and the suggested improvements !
